app/precompiled.py

```diff
@@ -713,8 +713,13 @@ def _extract_text_from_page(page, rect):
 
     def _get_address_from_get_textwords():
         text = page.get_text(clip=rect).strip()
-        # split and rejoin each line to remove any duplicate or trailing whitespace
-        return "\n".join(" ".join(line.split()) for line in text.split("\n"))
+        return "\n".join(
+            # split and rejoin each line to remove any duplicate or trailing whitespace
+            " ".join(line.split())
+            for line in text.split("\n")
+            # remove any lines that only contain empty strings
+            if line.strip()
```

Comment:
Not sure if it matters for this use case but it would be more robust to use `get_lines_with_normalised_whitespace`<sup>1</sup> from utils:

```suggestion
            line for line in get_lines_with_normalised_whitespace(text)
            # remove any lines that only contain empty strings
            if line.strip()
```

***

1. https://github.com/alphagov/notifications-utils/blob/4741402d194bb0dfc87d15b2dae16dd193cb5591/notifications_utils/formatters.py#L251C5-L252

Reply:
good thought - you made me realise i could rework this - because i dont really care what this function returns, as long as when we put it into a `PrecompiledPostalAddress` that address matches. So splitting words onto diff lines because they're in diff fonts is bad. trailing whitespace, who cares if we're deleting this

https://github.com/alphagov/notifications-template-preview/commit/47a8b163c724c8747615ce3b9fb1625adeab3a99